src/Workspaces/Core/Portable/Serialization/IsolatedAnalyzerReference.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#if NET
+
+using System;
+using System.Collections.Immutable;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Runtime.Loader;
+
+namespace Microsoft.CodeAnalysis.Serialization;
+
+/// <summary>
+/// Wrapper around a real <see cref="AnalyzerReference"/>.  An "isolated" analyzer reference is an analyzer reference
+/// associated with an <see cref="AssemblyLoadContext"/> that is connected to a set of other "isolated" analyzer
+/// references.  This allows for loading the analyzers and generators from it in a way that is associated with that load
+/// context, keeping them separate from other analyzers and generators loaded in other load contexts, while also
+/// allowing all of those instances to be collected when no longer needed.  Being isolated means that if any of the
+/// underlying assembly references change, that they can be loaded side by side with the prior references.  This enables
+/// functionality like live reloading of analyzers and generators when they change on disk.  Note: this is only
+/// supported on .Net Core, and not .Net Framework, as only the former has <see cref="AssemblyLoadContext"/>s.
+/// </summary>
+/// <remarks>
+/// The purpose of this type is to allow passing out a <see cref="AnalyzerReference"/> to the rest of the system that
+/// then ensures that as long as it is alive (or any <see cref="DiagnosticAnalyzer"/> or <see cref="ISourceGenerator"/>
+/// it passes out is alive), that the <see cref="IsolatedAssemblyReferenceSet"/> (and its corresponding <see
+/// cref="AssemblyLoadContext"/>) is kept alive as well.
+/// </remarks>
+internal sealed class IsolatedAnalyzerReference(
+    IsolatedAssemblyReferenceSet isolatedAssemblyReferenceSet,
+    AnalyzerReference underlyingAnalyzerReference) : AnalyzerReference
+{
+    /// <summary>
+    /// Conditional weak tables that ensure that as long as a particular <see cref="DiagnosticAnalyzer"/> or <see
```

Comment:
>Conditional weak tables [](http://example.com/codeflow?start=8&length=23)

Thinking out loud without yet understanding this code: 

Could we avoid the CWTs if we handed back wrapper DiagnosticAnalyzers that hold onto this object and delegate to the analyzer they wrap?

Reply:
it's a good question.  but the answer is a solid *no*.  The main reason for this is that many other parts of the system use teh actual type/instance of the DiagAnalyzer or ISourceGen to control logic.  For example, DiagAnalyzer is checked to see if it implements IBuiltInAnalyzer.  Similarly, we do have some controls depending on if the System.Type is from one of our assemblies or not.

For ISourceGen, later parts query to see if it's an IIncrementalSourceGen and stuff like that.  So actually wrapping these with outer types would almost certainly break expectations somewhere in the system.  The CWT is the safest way to ensure we pass back the actual instances, while keeping thing alive.